test/core/transport/chaotic_good/client_transport_test.cc

```diff
@@ -127,218 +244,447 @@ class ClientTransportTest : public ::testing::Test {
+  auto args = CallArgs{std::move(md),
+                       ClientInitialMetadataOutstandingToken::Empty(),
+                       nullptr,
+                       &pipe_server_intial_metadata_.sender,
+                       &pipe_client_to_server_messages_.receiver,
+                       &pipe_server_to_client_messages_.sender};
   StrictMock<MockFunction<void(absl::Status)>> on_done;
   EXPECT_CALL(on_done, Call(absl::OkStatus()));
   EXPECT_CALL(control_endpoint_, Write).WillOnce(Return(true));
   EXPECT_CALL(data_endpoint_, Write).WillOnce(Return(true));
   auto activity = MakeActivity(
       Seq(
-          // Concurrently: send message into the pipe, and receive from the
-          // pipe.
-          Join(Seq(pipe_client_to_server_messages_.sender.Push(
-                       std::move(messages[0])),
-                   [this] {
-                     this->pipe_client_to_server_messages_.sender.Close();
-                     return absl::OkStatus();
-                   }),
-               client_transport_.AddStream(std::move(args))),
+          // Concurrently: write and read messages in client transport.
+          Join(
+              // Send messages to call_args.client_to_server_messages pipe,
+              // which will be eventually sent to control/data endpoints.
+              Seq(pipe_client_to_server_messages_.sender.Push(
+                      std::move(messages[0])),
+                  [this] {
+                    pipe_client_to_server_messages_.sender.Close();
+                    return absl::OkStatus();
+                  }),
+              // Add first stream with call_args into client transport.
+              Seq(Join(client_transport_->AddStream(std::move(args)),
+                       [this]() {
+                         // Concurrently: start read from control endpoints.
+                         read_callback[0](absl::OkStatus());
+                         return absl::OkStatus();
+                       }),
+                  [](std::tuple<absl::StatusOr<ServerMetadataHandle>,
+                                absl::Status>
+                         ret) {
+                    // AddStream will finish with server trailers:
+                    // "grpc-status:0".
+                    EXPECT_EQ(std::get<0>(ret)
+                                  .value()
+                                  ->get(GrpcStatusMetadata())
+                                  .value(),
+                              grpc_status_code::GRPC_STATUS_OK);
+                    return absl::OkStatus();
+                  }),
+              // Receive messages from control/data endpoints.
+              Seq(
+                  // Receive server initial metadata.
+                  pipe_server_intial_metadata_.receiver.Next(),
+                  // Receive server to client messages.
+                  pipe_server_to_client_messages_.receiver.Next(),
+                  [this] {
+                    // Close pipes after receive message.
```

Comment:
Can you add expectations here and elsewhere to validate that the initial metadata and message values match the expected ones ?

Reply:
Sure, done.